feat: introduce driver/kernel yaml ignorelist

iurly opened the pull request:
introduce an external ignorelist, i.e. a way to specify a set of kernel releases that should be skipped for a given set of driver versions.
The set of releases is specified through a regex and a conditional expression on the exploded values, expressed as a Jinja2 template.
The most notable use case is an older driver version which is incompatible with newer kernels due to compiler issues.

iurly committed:
feat: introduce driver/kernel yaml blacklist

introduce an external blacklist, i.e. a way to specify a set of
kernel releases that should be skipped for a given set of
driver versions.
The set of releases is specified through a regex and a
conditional expression on the exploded values, expressed
as a Jinja2 template.
The most notable use case is an older driver version which
is incompatible with newer kernels due to compiler issues.

gnosek commented on probe_builder/builder/blacklist.py:
```diff
@@ -0,0 +1,86 @@
+        for b in self.blacklist:
+            logger.debug("==== evaluating blacklist entry {}".format(b['description']))
+            if self.probe_version in b['probe_versions']:
+                logger.debug("probe version {} matches one of agent versions {}".format(self.probe_version, b['probe_versions']))
```
This could be info level

iurly replied:
Hmm... not convinced. In the end this line will be executed for ALL kernels we'd building for that probe version. That's a little too much IMHO.

gnosek replied:
Oh. It looked like a log only for skipped kernels to me. Same for all the other logs, if they're emitted for all kernels, then yeah, that's too much, but I think we could use a log to tell us that a particular kernel is explicitly skipped because $REASON (I don't have a strong opinion on what log that is though).

iurly replied:
Actually such a log is already due to be printed at later point in time: https://github.com/draios/probe-builder/pull/112/files/593918521e7e1843dcaaac4cc7d41e7326302643#diff-de0cd1a9ef33f7f8110f51f64de81f61570a645378c421e59a8074c1892f02fbR89

My take is that one should be enough but if there's any other detail I'm missing, please shoot!

gnosek commented on probe_builder/builder/blacklist.py:
```diff
@@ -0,0 +1,86 @@
+                continue
+
+            if not b['probe_kinds'] or probe_kind in b['probe_kinds']:
+                logger.debug("probe kind {} matches one of {}".format(probe_kind, b['probe_kinds']))
```
Same, could be info level

iurly replied:
same here, isn't it too much?

gnosek commented on probe_builder/builder/blacklist.py:
```diff
@@ -0,0 +1,86 @@
+                t = jinja2.Template(sif)
+                r = t.render(m.groupdict())
+                if r and r != "False":
+                    logger.debug("== Blacklisted by {}, eval returned {}({})".format(b['description'], r, type(r)))
```
Same, could be info level

iurly replied:
well this one would indeed make sense, even though the same information is also available at the caller side

gnosek commented on probe_builder/builder/blacklist.py:
```diff
@@ -0,0 +1,86 @@
+        self.probe_version = probe_version
+
+    def blacklist_reason(self, probe_kind, kernel_release):
+        logger.debug("======== {} probe {} kernel {}".format(probe_kind, self.probe_version, kernel_release))
```
We could add a note that this is checking the blacklist, make this info level and drop the logs in build_kernel()

iurly replied:
not sure why you'd want to pollute the logs with so much verbosity

gnosek commented on probe_builder/builder/blacklist.py:
```diff
@@ -0,0 +1,86 @@
+    handler.setLevel(logging.DEBUG)
+    logger.addHandler(handler)
+    kbl = KernelBlacklist(_test_blacklist, "12.16.3")
+    assert kbl.blacklist_reason("kmod", "4.18.0-497.el8.x86_64")
+    assert not kbl.blacklist_reason("kmod", "4.18.0-506.el8.x86_64")
+    assert not kbl.blacklist_reason("kmod", "6.2.0")
+    kbl = KernelBlacklist(_test_blacklist, "12.12.0")
+    assert kbl.blacklist_reason("kmod", "6.2.0")
+    assert kbl.blacklist_reason("kmod", "7.0.0")
```
Ideally we'd check the exact reason as well (not that it's just not empty) to serve as documentation

iurly replied:
absolutely agree! done in follow-up commit

gnosek reviewed (dismissed):
Some nits about logging, but LGTM!

iurly committed:
add schema validation

mavimo commented on probe_builder/builder/blacklist.schema.json:
```diff
@@ -0,0 +1,66 @@
+    "matchers": {
+      "type": "object",
+      "description": "list of matches",
+      "default": []
```
```suggestion
      "patternProperties": {
        "^[a-z_]*$": {
          "type": "string",
        }
      },
      "default": []
```

iurly replied:
sorry, I missed your comments entirely. I will apply the suggestion in a follow-up commit.

iurly replied:
done in c193cfa

mavimo commented on probe_builder/builder/blacklist.py:
```diff
@@ -0,0 +1,86 @@
+
+
+## =========== Test code =============
+_test_blacklist = """\
```
Not sure to have test data as part of the application instead of having a test suite running out of the main application.

iurly replied:
you're right, but since there is no existing test infrastructure to hook up into I opted for the "quick and dirty" way to at least have a way to test (albeit manually)

mavimo commented on agent_blacklist.yaml:
```diff
@@ -0,0 +1,17 @@
+# Blacklist for known versions of sysdig agent (sysdigcloud-probe)
```
Is this file supposed to be here? 🤔

iurly replied:
It looked like the quickest way to get this going (and share it with customers if necessary, too)

iurly committed:
rename to ignorelist plus minor touches

- rename to ignorelist
- add .gitignore in github action
- add extra verification in assert statements
- fix 6.2 skip_if expression

iurly committed:
only allow id-like matchers

iurly committed:
Merge branch 'dev' into blacklist

aroberts87 commented on probe_builder/__init__.py:
```diff
@@ -156,7 +163,7 @@ def build(builder_image_prefix,
         for release, target in kernel_dirs:
             drel, krel = release if type(release) is tuple else ("", release)
-            future = executor.submit(distro_builder.build_kernel, workspace, probe, distro.builder_distro, krel, target)
+            future = executor.submit(distro_builder.build_kernel, kil, workspace,  probe, distro.builder_distro, krel, target)
```
```suggestion
            future = executor.submit(distro_builder.build_kernel, kil, workspace, probe, distro.builder_distro, krel, target)
```

iurly replied:
done in cc847ff

aroberts87 commented on probe_builder/builder/ignorelist.py:
```diff
@@ -0,0 +1,87 @@
+            config = yaml.safe_load(yamldoc)
+            for mn, mv in config['matchers'].items():
+                mv = re.compile(mv)
+                self.matchers[mn] = mv
```
`self.matchers = {mn: re.compile(mv) for mn, mv in config['matchers'].items()}`
☝🏻 this reads a bit easier to me, but you don't need to change it.

iurly replied:
done in cc847ff

aroberts87 commented on probe_builder/builder/ignorelist.py:
```diff
@@ -0,0 +1,87 @@
+    def ignore_reason(self, probe_kind, kernel_release):
+        logger.debug("======== {} probe {} kernel {}".format(probe_kind, self.probe_version, kernel_release))
+        for b in self.ignorelist:
+            logger.debug("==== evaluating ignorelist entry {}".format(b['description']))
```
Wouldn't we want to surround the description in quotes or something to distinguish it from the rest of the message?

iurly replied:
done in cc847ff

aroberts87 reviewed (dismissed):
lgtm. a few minor notes that you can decide to take or leave. 🙂

gnosek commented on probe_builder/builder/distro/base_builder.py:
```diff
@@ -108,12 +108,20 @@ def build_kernel_impl(cls, config_hash, container_name, image_name, kernel_dir,
 
         kmod_skip_reason = builder_image.skip_build(workspace.machine, probe, output_dir, release, config_hash, False)
+        if not kmod_skip_reason:
+            logger.info("Querying kmod ignorelist for {}".format(release))
```
This looks like it's going to happen for every kernel, right? I was thinking about replacing it (and the eBPF one below) with the slightly more informative log in ignorelist.py:22

iurly replied:
Yes, this is printed for every kernel. I used it essentially for debugging the change, so it should be more of a `logger.debug()` in my opinion.

iurly replied:
converted to `logger.debug()` in cc847ff

iurly committed:
apply suggestions from code review

- cosmetic changes/refactoring
- convert querying log from info to debug

deepskyblue86 commented on probe_builder/__init__.py:
```diff
@@ -130,11 +130,12 @@ def prebuild(builder_image_prefix, machine):
 @click.option('-t', '--download-timeout', type=click.FLOAT)
 @click.option('-v', '--probe-version')
 @click.option('-m', '--machine', default=os.uname().machine)
+@click.option('-l', '--kernel_ignorelist', default='')
```
All the options are with dash, with this we'll have
- `--kernel-filter`
- `--kernel-type`
... and `--kernel_ignorelist`

iurly replied:
renamed as `--ignore-list`

deepskyblue86 commented on probe_builder/builder/ignorelist.py:
```diff
@@ -0,0 +1,85 @@
+        self.ignorelist = []
+        if yamldoc:
+            config = yaml.safe_load(yamldoc)
+            self.matchers = { mn: re.compile(mv) for mn, mv in config['matchers'].items() }
+            self.ignorelist = config['ignorelist']
```
Is it fine to rely on just the validator, or better to check the config keys here?

deepskyblue86 commented on probe_builder/builder/ignorelist.py:
```diff
@@ -0,0 +1,85 @@
+                logger.debug("probe version {} NOT in {}".format(self.probe_version, b['probe_versions']))
+                continue
+
+            if not b['probe_kinds'] or probe_kind in b['probe_kinds']:
```
So no probe_kinds means any kind will be considered for the ignorelist, right?

iurly replied:
correct

deepskyblue86 commented on probe_builder/builder/ignorelist.py:
```diff
@@ -0,0 +1,85 @@
+
+            m = self.matchers[b['matcher']].search(kernel_release)
+            if m:
+                logger.debug("kernel {} matches matcher {}".format(kernel_release, b['matcher']))
```
is the kernel matching the matcher, or the matcher matching the kernel?

iurly replied:
that's a very good question... which I honestly don't want to answer

deepskyblue86 reviewed (dismissed):
Nice!
I think it would be useful to have a brief of your design doc as a `.md` file.

iurly committed:
rename --kernel_ignorelist to --ignore-list

iurly committed:
refine conditions for ignore kernels

iurly committed:
Merge branch 'dev' into blacklist

francesco-racciatti commented on probe_builder/builder/ignorelist.py:
```diff
@@ -0,0 +1,86 @@
+        return None
+
+
+## =========== Test code =============
+_test_ignorelist = """\
+matchers:
+  redhat: ^(?P<version>[0-9]\.[0-9]+\.[0-9]+)-(?P<rpmrelver>[0-9]+)(?P<rpmrelpatch>(\.[0-9]+)*)\.(?P<rhel>el.*)\.(?P<arch>[a-z0-9-_]+)$
+  generic: ^(?P<major>[0-9])\.(?P<minor>[0-9]+)\..*
+
+ignorelist:
+  - description: "492<=rhel<=500"
+    probe_versions: [ 12.16.0, 12.16.1, 12.16.2, 12.16.3, 12.17.0 ]
+    probe_kinds: [ kmod ]
+    matcher: redhat
+    skip_if: "{{ (version == '4.18.0' and 492<=(rpmrelver|int)<=500) }}"
+
+  - description: "kernel>=6.2"
+    probe_versions: [ 12.12.0 ]
+    probe_kinds: [ kmod ]
+    matcher: generic
+    skip_if: "{{ (major|int > 6) or (major|int == 6 and minor|int >= 2) }}"
+"""
+
+if __name__ == "__main__":
+    logger.setLevel(logging.DEBUG)
+    handler = logging.StreamHandler(sys.stderr)
+    handler.setFormatter(logging.Formatter('%(asctime)s %(message)s'))
+    handler.setLevel(logging.DEBUG)
+    logger.addHandler(handler)
+    kil = KernelIgnoreList(_test_ignorelist, "12.16.3")
+    assert "492<=rhel<=500" == kil.ignore_reason("kmod", "4.18.0-497.el8.x86_64")
+    assert None == kil.ignore_reason("kmod", "4.18.0-477.27.1.el8_8.x86_64")
+    assert None == kil.ignore_reason("kmod", "4.18.0-506.el8.x86_64")
+    assert None == kil.ignore_reason("kmod", "6.2.0")
+    kil = KernelIgnoreList(_test_ignorelist, "12.12.0")
+    assert "kernel>=6.2" == kil.ignore_reason("kmod", "6.2.0")
+    assert "kernel>=6.2" == kil.ignore_reason("kmod", "7.0.0")
+    assert None == kil.ignore_reason("kmod", "6.1.99")
```
🙂

francesco-racciatti reviewed (approved):
lgtm
